compiler/rustc_passes/src/dead.rs

```diff
@@ -1113,10 +1166,13 @@ fn check_mod_deathness(tcx: TyCtxt<'_>, module: LocalModDefId) {
             || (def_kind == DefKind::Trait && live_symbols.contains(&item.owner_id.def_id))
         {
             for &def_id in tcx.associated_item_def_ids(item.owner_id.def_id) {
-                // We have diagnosed unused methods in traits
+                // We have diagnosed unused assoc consts and fns in traits
                 if matches!(def_kind, DefKind::Impl { of_trait: true })
-                    && tcx.def_kind(def_id) == DefKind::AssocFn
-                    || def_kind == DefKind::Trait && tcx.def_kind(def_id) != DefKind::AssocFn
+                    && matches!(tcx.def_kind(def_id), DefKind::AssocConst | DefKind::AssocFn)
+                    // skip unused public inherent methods,
+                    // cause we have diagnosed unconstructed struct
+                    || matches!(def_kind, DefKind::Impl { of_trait: false }) && tcx.visibility(def_id).is_public() && ty_ref_to_pub_struct(tcx, tcx.hir().item(item).expect_impl().self_ty)
+                    || def_kind == DefKind::Trait && tcx.def_kind(def_id) == DefKind::AssocTy
```

Comment:
I think at current, even if the structure has never been constructed, the inherent methods will still check whether they are used or not?

Reply:
Only public inherent methods without the receiver would be checked firstly. If no such methods and the struct is not constructed by other private methods, then rest public inherent methods won't be checked.

For private inherent methods, they only be checked when then are called.

Reply:
That's ok